VSRAD.Package/DebugVisualizer/Wavemap/WavemapImage.cs

```diff
@@ -0,0 +1,169 @@
+
+        private int _xOffset = 0;
+        public int XOffset
+        {
+            get => _xOffset;
+            set
+            {
+                _xOffset = value;
+                SetData(_view);
+            }
+        }
+
+        private int _yOffset = 0;
+        public int YOffset
+        {
+            get => _yOffset;
+            set
+            {
+                _yOffset = value;
+                SetData(_view);
+            }
+        }
+
+        private int _elementsWidthX = 100;
+        private int _elementsWidthY = 8;
+
+
+        public WavemapImage(PictureBox box, VisualizerContext context)
+        {
+            _box = box;
+            _context = context;
+        }
+
+        private void ShowWaveInfo(object sender, MouseEventArgs e)
+        {
+            var row = (e.Y / _rSize) + _elementsWidthY * _yOffset;
+            var col = (e.X / _rSize) + _elementsWidthX * _xOffset;
+            var waveInfo = _view[row, col];
+            _context.CurrentWaveInfo = waveInfo.IsVisible
+                ? $"G: {col}, W: {row}, L: {waveInfo.BreakLine}"
+                : $"G: {col}, W: {row}, Out of range.";
+        }
+
+        public void SetData(WavemapView view)
+        {
+            if (view == null || view.WavesPerGroup == 0)
+            {
+                _box.Image = null;
+                return;
+            }
+
+            _view = view;
+            var pixelCount = _elementsWidthX * _elementsWidthY * (_rSize + 1) * (_rSize + 1);
+            var byteCount = pixelCount * 4;
+            var imageData = new byte[byteCount + 54];
+            _header.CopyTo(imageData, 0);
+            var fileSizeBytes = BitConverter.GetBytes(54 + byteCount);
+            var widthBytes = BitConverter.GetBytes(_elementsWidthX * _rSize + 1);
+            var heightBytes = BitConverter.GetBytes(_elementsWidthY * _rSize + 1);
+            var dataSizeBytes = BitConverter.GetBytes(byteCount);
+            for (int i = 0; i < 4; i++)
+            {
+                imageData[2 + i] = fileSizeBytes[i];
+                imageData[18 + i] = widthBytes[i];
+                imageData[22 + i] = heightBytes[i];
+                imageData[34 + i] = dataSizeBytes[i];
+            }
+            var byteWidth = _elementsWidthX * _rSize * 4 + 4;   // +4 for right border
+
+            for (int i = 0; i < byteCount - 3; i += 4)
+            {
+                int row = i / byteWidth;
+                if ((row % _rSize) == 0 || row / _rSize >= _elementsWidthY) continue;
+                int col = i % byteWidth;
+                if ((col % _rSize) == 0) continue;
+
+                var viewRow = (_elementsWidthY - 1 - row / _rSize) + _elementsWidthY * _yOffset;
+                var viewCol = (col / _rSize / 4) + _elementsWidthX * _xOffset;
+                var waveInfo = view[viewRow, viewCol];
+
+                var flatIdx = i + 54;   // header offset
+
+                imageData[flatIdx+0] = waveInfo.BreakColor[0]; // B
```

Comment:
The new implementation is much more responsive than the previous one, and the time spent in `SetData` is insignificant compared to what we waste in `DataGridView`, but it still wouldn't hurt to avoid some basic performance pitfalls. Rough measurements show that this loop can take up to 12ms. `GetWaveInfoByRowAndColumn` (`view[viewRow, viewCol];`) is particularly expensive and is invoked for every non-border pixel. I believe you can reduce the number of calls to it by a factor of `_rSize` with a simple loop over each rectangle line:

```
var flatIdx = i + 54;   // header offset

for (int rwidth = 0; rwidth < _rSize - 1; ++rwidth)
{
    imageData[flatIdx + 0] = waveInfo.BreakColor[0]; // B
    imageData[flatIdx + 1] = waveInfo.BreakColor[1]; // G
    imageData[flatIdx + 2] = waveInfo.BreakColor[2]; // R
    imageData[flatIdx + 3] = waveInfo.BreakColor[3]; // Alpha
    flatIdx += 4;
}
i += (_rSize - 2) * 4;
```